odm/firestore_abstract_repository.go

```diff
@@ -0,0 +1,252 @@
+package odm
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/firestore"
+	"github.com/SaiNageswarS/go-api-boot/logger"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.uber.org/zap"
+	"google.golang.org/api/iterator"
+)
+
+type FirestoreBootRepository[T any] interface {
+	Save(model DbModel) chan error
+	FindOneById(id string) (chan *T, chan error)
+	IsExistsById(docID string) bool
+	CountDocuments(query firestore.Query) (chan int64, chan error)
+	FindOne(filters bson.M) (chan *T, chan error)
+	Find(filters bson.M, sort bson.D, limit, skip int) (chan []T, chan error)
+	DeleteById(id string) chan error
+	DeleteOne(filters bson.M) chan error
+}
+
+type UnimplementedFirestoreBootRepository[T any] struct {
+	Database       string
+	CollectionName string
+}
+
+func (r *UnimplementedFirestoreBootRepository[T]) db() *firestore.Client {
+	return GetFireStoreClient(r.Database)
+}
+
+func (r *UnimplementedFirestoreBootRepository[T]) Save(model DbModel) chan error {
+	ch := make(chan error)
+	ctx := context.Background()
+	go func() {
+		id := model.Id()
+		collection := r.db().Collection(r.CollectionName)
+		documentRef := collection.Doc(id)
+		document := make(map[string]interface{})
+		if r.IsExistsById(id) {
+			document["updatedOn"] = time.Now().Unix()
+		} else {
+			document["createdOn"] = time.Now().Unix()
+		}
+		documentRef.Set(ctx, model)
+		documentRef.Set(ctx, document, firestore.MergeAll)
+		ch <- nil
+	}()
+
+	return ch
+}
+
+func (r *UnimplementedFirestoreBootRepository[T]) IsExistsById(docID string) bool {
+	resCh := make(chan bool)
+	go func() {
+		docRef := r.db().Collection(r.CollectionName).Doc(docID)
+		docSnap, err := docRef.Get(context.Background())
+		if err != nil {
+			fmt.Println("hit id existanc eerro")
+			logger.Error("error checking document existence", zap.Error(err))
+			resCh <- false
+			return
+		}
+		if docSnap.Exists() {
+			resCh <- true
+			return
+		}
+	}()
+	fmt.Println("request processed")
+	res := <-resCh
+	return res
+}
+
+func (r *UnimplementedFirestoreBootRepository[T]) CountDocuments(query firestore.Query) (chan int64, chan error) {
+	countChan := make(chan int64)
+	errChan := make(chan error)
+
+	go func() {
+		ctx := context.Background()
+		documents, err := query.Documents(ctx).GetAll()
+
+		if err != nil {
+			errChan <- err
+			return
+		}
+
+		countChan <- int64(len(documents))
+	}()
+
+	return countChan, errChan
+}
+
+func (r *UnimplementedFirestoreBootRepository[T]) Find(filters bson.M, sort bson.D, limit, skip int) (chan []T, chan error) {

```

Comment:
Let's make Find generic and change in both mongo and here to have filters and sort as json, for now, conforming to bson format.